File: mobile/stores/dailyLogStore.ts
import { create } from 'zustand';
import type { FoodEntry, MealLabel, Macros } from '@shared/types';
import * as api from '@/services/api';

function groupByMeal(entries: FoodEntry[]): Record<MealLabel, FoodEntry[]> {
  const groups: Record<MealLabel, FoodEntry[]> = {
    breakfast: [],
    lunch: [],
    dinner: [],
    snack: [],
  };
  for (const entry of entries) {
    groups[entry.mealLabel].push(entry);
  }
  return groups;
}

function computeTotals(entries: FoodEntry[]): Macros {
  return entries.reduce(
    (acc, e) => ({
      calories: acc.calories + e.calories,
      proteinG: acc.proteinG + e.proteinG,
      carbsG: acc.carbsG + e.carbsG,
      fatG: acc.fatG + e.fatG,
    }),
    { calories: 0, proteinG: 0, carbsG: 0, fatG: 0 },
  );
}

function recalculate(entries: FoodEntry[]) {
  return {
    entries,
    entriesByMeal: groupByMeal(entries),
    totals: computeTotals(entries),
  };
}

interface DailyLogState {
  entries: FoodEntry[];
  entriesByMeal: Record<MealLabel, FoodEntry[]>;
  totals: Macros;
  isLoading: boolean;
  error: string | null;
  fetch: (date: string) => Promise<void>;
  addEntry: (entry: FoodEntry) => void;
  removeEntry: (id: string) => FoodEntry | undefined;
  restoreEntry: (entry: FoodEntry) => void;
  commitDelete: (id: string) => Promise<void>;
}

export const useDailyLogStore = create<DailyLogState>((set, get) => ({
  entries: [],
  entriesByMeal: { breakfast: [], lunch: [], dinner: [], snack: [] },
  totals: { calories: 0, proteinG: 0, carbsG: 0, fatG: 0 },
  isLoading: false,
  error: null,

  fetch: async (date: string) => {
    set({ isLoading: true, error: null });
    try {
      const entries = await api.getEntries(date);
      set({ ...recalculate(entries), isLoading: false });
    } catch (e) {
      set({
        isLoading: false,
        error: e instanceof Error ? e.message : 'Failed to load entries',
      });
    }
  },

  addEntry: (entry: FoodEntry) => {
    const { entries } = get();
    const updated = [...entries, entry].sort(
      (a, b) =>
        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
    );
    set(recalculate(updated));
  },

  removeEntry: (id: string) => {
    const { entries } = get();
    const removed = entries.find((e) => e.id === id);
    if (!removed) return undefined;
    const remaining = entries.filter((e) => e.id !== id);
    set(recalculate(remaining));
    return removed;
  },

  restoreEntry: (entry: FoodEntry) => {
    const { entries } = get();
    const updated = [...entries, entry].sort(
      (a, b) =>
        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
    );
    set(recalculate(updated));
  },

  commitDelete: async (id: string) => {
    await api.deleteEntry(id);
  },
}));

File: MacroTrackerSwift/MacroTrackerSwift/Stores/DailyLogStore.swift
import Foundation
import Observation

/// Manages the food entries for the selected date.
/// Direct port of mobile/stores/dailyLogStore.ts.
@Observable @MainActor
final class DailyLogStore {
    static let shared = DailyLogStore()

    var entries:     [FoodEntry] = []
    var isLoading:   Bool = false
    var error:       String? = nil

    private init() {}

    // MARK: - Computed

    var entriesByMeal: [MealLabel: [FoodEntry]] {
        var groups: [MealLabel: [FoodEntry]] = [
            .breakfast: [], .lunch: [], .dinner: [], .snack: []
        ]
        for entry in entries {
            groups[entry.mealLabel, default: []].append(entry)
        }
        return groups
    }

    var totals: Macros {
        entries.reduce(.zero) { acc, e in
            Macros(
                calories: acc.calories + e.calories,
                proteinG: acc.proteinG + e.proteinG,
                carbsG:   acc.carbsG   + e.carbsG,
                fatG:     acc.fatG     + e.fatG)
        }
    }

    // MARK: - Actions

    func fetch(date: String) async {
        isLoading = true
        error = nil
        do {
            let fetched = try await APIClient.shared.getEntries(date: date)
            entries   = fetched.sorted { $0.createdAt < $1.createdAt }
            isLoading = false
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }

    func addEntry(_ entry: FoodEntry) {
        entries = (entries + [entry]).sorted { $0.createdAt < $1.createdAt }
    }

    @discardableResult
    func removeEntry(id: String) -> FoodEntry? {
        guard let idx = entries.firstIndex(where: { $0.id == id }) else { return nil }
        let removed = entries[idx]
        entries.remove(at: idx)
        return removed
    }

    func restoreEntry(_ entry: FoodEntry) {
        entries = (entries + [entry]).sorted { $0.createdAt < $1.createdAt }
    }

    func commitDelete(id: String) async throws {
        try await APIClient.shared.deleteEntry(id: id)
    }

    /// Creates an entry via the API, inserts it locally, and returns it.
    /// Throws on network failure so callers can handle undo / error feedback.
    @discardableResult
    func createEntry(_ request: CreateFoodEntryRequest) async throws -> FoodEntry {
        let entry = try await APIClient.shared.createEntry(request)
        addEntry(entry)
        return entry
    }

    // MARK: - Frequent / Recent Foods

    var frequentFoods: [FrequentFood] = []

    func fetchFrequentFoods(limit: Int = 3) async {
        do {
            let foods = try await APIClient.shared.getFrequentFoods()
            frequentFoods = Array(foods.prefix(limit))
        } catch {
            // non-critical — swallow silently
        }
    }
}
